### PR TITLE
Python package reorg

### LINKED ISSUE #4163: genf90 should be incorporated into CIME via submodule

### PR DESCRIPTION
**Pull request by jasonb5**
The goal of this PR is to consolidate all files required to run CIME 
under a single top-level directory. This updates CIME's structure to 
be more like a standardized python library.

**New**
- CIME/scripts
- CIME/data
  - Captures configuration/templates
- CIME/non_py
  - Contains non python code
- CIME/non_py/externals/genf90 is now a submodule
- tools/README.md

**Moved**
- config moved to CIME/data/config
- scripts/lib/CIME moved to CIME
- scripts/Tools moved to CIME/Tools
- src/build_scripts moved to CIME/build_scripts
- src/CMake moved to CIME/non_py/src/CMake
- src/components moved to CIME/non_py/src/components
- src/share/timings moved to CIME/non_py/src/timings
- src/eternals moved to CIME/non_py/externals
- tools/configure moved to CIME/scripts/configure
- tools/cprnc moved to CIME/non_py/cprnc
- scripts/lib/*.template moved to CIME/data/templates
- scripts/lib/get_tests.py moved to CIME/get_tests.py
- scripts/lib/jenkins_generic_job.py moved to CIME/jenkins_generic_job.py

**Removed**
- scripts/lib/six.py
- scripts/lib/six_additions.py

Test suite: scripts_regression_tests.py
Test baseline: n/a
Test namelist changes: n/a
Test status: n/a

Fixes #4163
User interface changes?: n/a
Update gh-pages html (Y/N)?: n

### CODE REVIEW

**Comment by jasonb5** [comment]
This update will most likely require external updates for models using CIME.

Reviewers let me know if I missed any files or if files should belong in different locations.

File: .github/workflows/srt.yml [hunk 1]
```diff
@@ -173,7 +173,7 @@ jobs:
           export NETCDF=$HOME/netcdf-fortran
           export PATH=$NETCDF/bin:$PATH
           export LD_LIBRARY_PATH=$NETCDF/lib:$HOME/pnetcdf/lib:$LD_LIBRARY_PATH
-          python -m pip install pytest pytest-cov
+          python -m pip install pytest pytest-cov six
```
Comment (jedwards4b) [inline .github/workflows/srt.yml:176]: We should no longer need six as we no longer support python 2.

File: CIME/BuildTools/macrowriterbase.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 from abc import ABCMeta, abstractmethod
 from CIME.XML.standard_module_setup import *
 from CIME.utils import get_cime_root
-from six import add_metaclass
+from CIME.six import add_metaclass
```
Comment (jedwards4b) [inline CIME/BuildTools/macrowriterbase.py:16]: let's just remove the need for six

File: CIME/Tools/Makefile [hunk 3]
```diff
@@ -967,7 +967,7 @@ $(COMPLIB): $(OBJS)
 	$(CXX) -c $(INCLDIR) $(INCS) $(CXXFLAGS)  $<
 
 %.F90: %.F90.in
-	$(CIMEROOT)/src/externals/genf90/genf90.pl $< > $@
+	$(CIMEROOT)/CIME/data/genf90/genf90.pl $< > $@
```
Comment (jedwards4b) [inline CIME/Tools/Makefile:970]: why is this here?

**Comment by billsacks** [comment]
Thank you very much for your work on this, @jasonb5! I haven't looked super closely at this (and I'm not sure if I will) but have a couple of thoughts from your list of moved things... by the way, thank you very much for giving that concise summary of what was moved – that is much easier to review!

Overall, this seems like a definite improvement in the organization, both in making this more like a standard python package and more generally. So thank you!

My main questions concern the CIME/data directory. Maybe it's just that it changes what I'm used to, but this doesn't feel like the most intuitive home for some of the things that are now in there. For some of these things, I'm not entirely sure what the right answer is (e.g., CIME/data/config, CIME/data/CMake), but some really feel to me like they should be elsewhere:
- CIME/data/components – is this actually still needed? (@fischer-ncar - this just includes some nuopc components; were these not split out in the CIME split?) If still needed, I think it should be in a directory at the top level, not nested under CIME/data, since it is really separate from the CIME python library.
- CIME/data/share – similar comments as for CIME/data/components
- CIME/data/cprnc – I feel it's more intuitive to keep this in a top-level tools directory. I realize there is some confusion because there were two different tools directories (scripts/Tools and tools). Maybe we should rename one of them.
- CIME/data/genf90 – Is this still an external, managed via git subtree? If so, it seems like there would still be value in keeping it in a directory that marks it as an external. If we reinstate a top-level tools directory (which I think we should), then it could go there (tools/externals/genf90)

Also, CIME/configure seems out of place, because it is the only(?) executable at the top level of CIME. Just like for cprnc, I liked this in a top-level tools directory.

**Review by jedwards4b — COMMENTED** [review]
Let's just remove the six package completely - you've already started doing it.

File: .github/workflows/srt.yml (hunk 1 above)
Comment (jasonb5) [reply .github/workflows/srt.yml:176]: > Let's just remove the six package completely - you've already started doing it.

Sounds good, I'll remove and note it in the PR description.

**Comment by jasonb5** [comment]
@billsacks The idea was to move everything required to run CIME under the CIME directory. There were two motivations here; one was to get rid of the relative paths that back out of the top-level package directory (more cosmetic that anything), the second was if we ever package this as a wheel (pypi or conda) then we need any "data" (configuration or non python files) to be under the CIME directory.

If we never plan on doing the latter, then I'd be fine moving things back under externals, src, etc.

You have a point about `CIME/configure` but I think it might make more sense under CIME/scripts since it's a script and I'd like to keep the python code under the CIME directory. 

If you find anything else that's deprecated or no longer used let me know and I can remove it.

File: CIME/Tools/Makefile (hunk 3 above)
Comment (jasonb5) [reply CIME/Tools/Makefile:970]: Is genf90 no longer used or supplied externally. I believe I found references in E3SM so I kept it. @jgfouca do we still used this?

Comment (jedwards4b) [reply CIME/Tools/Makefile:970]: genf90 is both used and supplied externally https://github.com/PARALLELIO/genf90

**Comment by fischer-ncar** [comment]
@billsacks CIME/data/components was still needed at the time I did the cime split.  I don't know if that's still true.

**Comment by jedwards4b** [comment]
I think that data/components/stubs is no longer required but data/components/xcpl is.

**Comment by billsacks** [comment]
> The idea was to move everything required to run CIME under the CIME directory. There were two motivations here; one was to get rid of the relative paths that back out of the top-level package directory (more cosmetic that anything), the second was if we ever package this as a wheel (pypi or conda) then we need any "data" (configuration or non python files) to be under the CIME directory.

Thanks for the explanation - that makes sense. (Sorry, you probably explained this before and I forgot. Also, sorry for missing a possible reminder of this at last Wednesday's meeting.)

So then I think the main thing is having these in subdirectories called something other than data, because "data" feels like a confusing place to put (Fortran) source code and tools.

What about putting components and share under a new directory like CIME/fortran_src, moving cprnc into CIME/Tools (maybe there's no need to distinguish between the two flavors of tools that we used to have), and moving genf90 into CIME/Tools/externals?

> You have a point about CIME/configure but I think it might make more sense under CIME/scripts since it's a script and I'd like to keep the python code under the CIME directory.

I like the idea of putting this in CIME/scripts. Like some of the other things in there, can we also put an executable version in the top-level scripts, so that configure would live along-side create_newcase, etc.? That feels like an intuitive place to me.

**Comment by jasonb5** [comment]
> I think that data/components/stubs is no longer required but data/components/xcpl is.

I'll see if E3SM references it at all.

File: CIME/Tools/Makefile (hunk 3 above)
Comment (jgfouca) [reply CIME/Tools/Makefile:970]: @jedwards4b , @jasonb5 , genf90 is used frequently in our build system. I'm wondering if we should bring it in as a submodule instead of having a copy directly in our CIME repo?

Comment (jedwards4b) [reply CIME/Tools/Makefile:970]: I agree - it should be a submodule.

Comment (jasonb5) [reply CIME/Tools/Makefile:970]: @jgfouca I can add it as a submodule.

Comment (rljacob) [reply CIME/Tools/Makefile:970]: So genf90 is no longer needed by cime itself?

Comment (jedwards4b) [reply CIME/Tools/Makefile:970]: That's correct, all of the F90 code has been removed from cime.

Comment (jgfouca) [reply CIME/Tools/Makefile:970]: @rljacob , it is still used. I see it being used in the key makefile `scripts/Tools/Makefile` . It also looks like some sharedlibs and tools use it (cprnc and pio).

**Comment by jasonb5** [comment]
This is ready to start being reviewed. The last updates organizes `CIME/data` into more logical sub-directories and brings `genf90` in as a submodule.

I'll be adding some documentation to note that `git submodule update --init` should be ran after clone/pulls.

This needs to be pretty thoroughly tested to ensure I didn't miss any files.

**Comment by jedwards4b** [comment]
@jasonb5 can you give me permission to push back to your fork?

**Comment by jasonb5** [comment]
@jedwards4b Should have invite.

**Review by jgfouca — APPROVED** [review]

**Comment by rljacob** [comment]
E3SM has its own versions of xcpl and stubs for building cases with those pieces.

Do we still want a clone of ESMCI/cime by itself to be able to build and run something?

**Comment by rljacob** [comment]
Did we come up with a better name for the directory-of-non-python-things besides "data"?

**Comment by jasonb5** [comment]
I left the xcpl and stub components as they're still referenced in the `config_files.xml` for `ufs` and `cesm` nuopc.

I think we should still have something that builds and runs. I image at some point we'll have moved all the model specific configuration out of `cime` and should have something we can build and run that is model independent.

I haven't heard anything yet for a possible replacement for `data`.

**Comment by jedwards4b** [comment]
Note that if you want to see the git log for code previous to moving the files you need to use
```
git log --follow filename
```
this only works for a single file at a time.

**Comment by jedwards4b** [comment]
Running into a problem here:

**Comment by jedwards4b** [comment]
I've pushed branches of cp7
```
remote:      https://github.com/jedwards4b/CESM_CPL7andDataComps/pull/new/cime_python_reorder
remote:
To github.com:jedwards4b/CESM_CPL7andDataComps.git
 * [new branch]      cime_python_reorder -> cime_python_reorder
```
and ccs_config:
```
remote:      https://github.com/jedwards4b/ccs_config_cesm/pull/new/cime_python_reorder
remote:
To github.com:jedwards4b/ccs_config_cesm.git
 * [new branch]      cime_python_reorder -> cime_python_reorder
```
to support the cime python reorder.    I have also created a branch from that PR:
```
remote:      https://github.com/jedwards4b/cime/pull/new/python_package_reorg_jpe
remote:
To github.com:/jedwards4b/cime.git
 * [new branch]          python_package_reorg_jpe -> python_package_reorg_jpe
```
The ccs_config changes are needed in any case, the changes to cpl7 and cime are only needed to avoid the error in
xmllint that I still don't understand.

**Comment by billsacks** [comment]
Thanks for all of your continued work on this @jasonb5, and thanks also @jedwards4b for the work you have been doing on this.

Regarding the naming of the data subdirectory: @jasonb5, sorry, until now I never saw your comment:

> I would like to keep all the non-python code under a single folder, I agree with improving the organization under the folder. I think data works for non-python code but I'm open to other suggestions.

(It looks like you made that comment in an edit. GitHub doesn't send notifications for edits, so I never saw that until I was skimming back through earlier comments just now.)

Can you explain your thinking for why it's important to have this all under a single folder? If we could have two folders, I'd suggest `data` and something like `non_python_src`. If we have to put it under a single folder, then I'm struggling to come up with a name, but I would prefer something generic but accurate like `non_python` rather than the more specific but misleading (to me anyway) `data`.

> I'll move the configure file under scripts. I'm probably missing some information but is the configure used outside of CIME (e.g. does the user ever call it manually)?

Yes, potentially, if I remember correctly: This utility is intended for use by other tool builds, so that you can build some other standalone tool using the CIME build scripts. That said, I can't remember if it is still used in this way, but I think that was its intent.

**Comment by jasonb5** [comment]
@billsacks
The thought was just purely organizational, the idea of having one directory to find all non-python files. If it makes more sense I can dump the `data` directory and place everything directory under the `CIME` directory.

External tools can still use the configure script they just need to point to `CIME/scripts`. The thought behind the root `scripts` directory was to capture scripts called directly by users. If external scripts need any tools they should point to the `CIME` directory.

If the use of the root `scripts` directory is fine, I should finish clean it out, there are test files that should be moved under `CIME/tests`.

**Comment by billsacks** [comment]
Thanks @jasonb5 . In terms of the data directory, I feel like I have voiced my preferences, but I'm willing to go with anything that makes sense to others, so at this point I'd like to hear what others think. I do like your idea of not having *too many* different subdirectories of non-python code, but I'm suggesting two rather than one: one for things that are really data and one for things that are non-python source code (tools, etc.).

**Comment by billsacks** [comment]
> I'm suggesting two rather than one: one for things that are really data and one for things that are non-python source code (tools, etc.)

or one that is very generically named

**Comment by jasonb5** [comment]
@billsacks @rljacob 
Most projects I've seen that provide **non-python data** just provide it under the package directory. I've seen this **data** organized in a variety of ways but it's almost always under the python package directory. The projects I've seen with **non-python code** usually store it outside the python package but compile it into a shared library and distribute that not the actual code.

**Comment by gold2718** [comment]
It would be great if this reorganization made it easier to find tools that are distributed with CIME. Right now, there is:
- utils
- tools (with its ten subdirectories)
- cime/scripts/Tools

I vote for a single "non_python" directory (which can be anywhere in the tree). Having "data" and "tools" under there would be great.

But what happens to python scripts that are not directly tied to the CIME CCS (e.g., find_circular_dependency.py)? Where will they end up?

**Comment by rljacob** [comment]
Note from telecon:
CIME/data will just have "templates" and "config" in it
new dir CIME/nopy will have "externals",  "src" and "cprnc".

**Comment by billsacks** [comment]
Did we agree on `nopy`? I thought it was going to be something like `non_py`, which is easier to make sense of.

**Comment by rljacob** [comment]
I don't remember.  non_py is fine.

Some more suggestions:
Since perl5lib is the only thing in utils, bring it up one (remove "utils")
Put a README in tools that describes what is there and points to where cprnc moved to.

**Comment by jasonb5** [comment]
I'm fine with non_py as well.

These additional suggestions sound great. One thing to note is moving `perl5lib` will require updates to external components using it, as long as we're fine with that I can move it.

**Comment by jedwards4b** [comment]
@jasonb5 can you just create a link to make the move of perl5lib backward compatible?

**Comment by jasonb5** [comment]
@jedwards4b @rljacob Maybe we should leave it under `tools/perl5lib` then otherwise we'll have `perl5lib` and then symlink to `tools/perl5lib`.

**Comment by jasonb5** [comment]
@jedwards4b @rljacob Added the changes we discussed. Let me know if anything is missing from `tools/README.md`.

**Comment by rljacob** [comment]
cprnc isn't in that directory so the README should literally say "cprnc was located here but has moved to  (path)."  That's for the benefit of people who expect it there.  Newbies can go to that the new directory to read about it.  I'd also request no markup since most people reading this will be seeing it in a terminal.

**Comment by jasonb5** [comment]
@jedwards4b @mvertens Any update on testing this PR?

**Comment by mvertens** [comment]
@jasonb5 - @fischer-ncar has done testing and I think we are ready to go ahead with the merge. However, @jedwards4b is on PTO and will be back tomorrow. I'd like to make sure he okays things as well before we do the merge. Is that okay? Sorry for the delay in finalizing this. Thanks for all of your work on this big improvment.

**Comment by jasonb5** [comment]
@mvertens I'm fine waiting for @jedwards4b input. Glad to hear there was testing and no issues.

**Comment by mvertens** [comment]
@jasonb5 - @fischer-ncar did encounter a few issues but they were mostly with the mct cases and are not show stoppers. We can target fixing them after the merge - particularly since we are in the process of deprecating these cases.

**Review by jedwards4b — APPROVED** [review]

**Review by mvertens — APPROVED** [review]

**Comment by jedwards4b** [comment]
The goal of this PR is to consolidate all files required to run CIME
under a single top-level directory. This updates CIME's structure to
be more like a standardized python library.

New

    CIME/scripts
    CIME/data
        Captures configuration/templates
    CIME/non_py
        Contains non python code
    CIME/non_py/externals/genf90 is now a submodule
    tools/README.md

Moved

    config moved to CIME/data/config
    scripts/lib/CIME moved to CIME
    scripts/Tools moved to CIME/Tools
    src/build_scripts moved to CIME/build_scripts
    src/CMake moved to CIME/non_py/src/CMake
    src/components moved to CIME/non_py/src/components
    src/share/timings moved to CIME/non_py/src/timings
    src/eternals moved to CIME/non_py/externals
    tools/configure moved to CIME/scripts/configure
    tools/cprnc moved to CIME/non_py/cprnc
    scripts/lib/*.template moved to CIME/data/templates
    scripts/lib/get_tests.py moved to CIME/get_tests.py
    scripts/lib/jenkins_generic_job.py moved to CIME/jenkins_generic_job.py

Removed

    scripts/lib/six.py
    scripts/lib/six_additions.py

Test suite: scripts_regression_tests.py
Test baseline: n/a
Test namelist changes: n/a
Test status: n/a

Fixes https://github.com/ESMCI/cime/issues/4163
User interface changes?: n/a
Update gh-pages html (Y/N)?: n
